### DISCUSSION TITLE
Ability to use submodules for easy sharing of Actions + Workflows

### DISCUSSION THREAD

**Discussion by saudeon** [community/community#10892] [Actions]
👋🏻  
Hey GitHub Team!
First time writer, long time user. Figure it was about time I added to the overall discussion of GitHub features.

Basically, as the title suggests, I'd like to be able to utilize submodules to share our Actions + Workflows hidden within the `.github` directory. It'd be nice to be able to do it with the `.github` directory as well, but that's a different conversation/discussion entirely.

Unfortunately, I understand that due to a submodule being a file reference to a commit hash and not a directory itself, it seems that GitHub does not recognize the available actions and workflows within the submodule. It would be great to be able to do this, as that would enable [insert redacted company name here] to have multiple repositories (one per distributed library) with a single source of truth for actions and workflows to run.

I created a repository to showcase what I would like to achieve one day. It'd make our developer experience just that much better in the long run.

[Workflows and Actions as submodules](https://github.com/saudeon/submodule-workflow-action)

**Comment by ethomson** [comment]
Hi @saudeon - I must confess that I find git submodules a bit...  frustrating to use, and I would love to find a solution that is a little less...  difficult.  How much do [reusable workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows) help here, and what sort of pain points do you still have with them?

**Reply by saudeon** [reply]
Hi @ethomson - I admit, a lot of people have been burned by `git submodule` quite a bit in the time of utilizing `git`. I believe if setup appropriately, it shouldn't hinder the developers process. It would also just be in-addition to, not solely. See this as a potential solution in a myriad of other ways, what those other ways are, I'm unsure of.

Reusable workflows don't help us due to the fact that our codebases are completely private, and we would have distributed repositories not just working off of one. From my understanding, reusable workflows require the following.

> Both workflows are in the same repository.
> The called workflow is stored in a public repository.

This also doesn't help the fact that our actions are also private.

**Reply by chrispat** [reply]
@saudeon

Reusable workflows can be referenced across repositories and today they can be either public or internal.  Adding support for private repositories in the same organization is in progress and should ship in the not too distant future.

**Reply by backmeupplz** [reply]
@chrispat the reason we can't use it is that we will still need to add the workflow files referencing the reusable workflows to every repository. And the workflows aren't the only thing that we'd want to reuse — what about other github settings like the pull request template? 🤔

IMHO, we might want to be able to set `.github` as a submodule, but having the solution with reusable workflows pointing at the `@main` branch of the reusable workflow might suffice for now 👍

**Reply by saudeon** [reply]
Agreed with @backmeupplz, it’s a great first step but being able to use a git sub module to reuse the entire `.github` folder would be far more beneficial, as we can then supply all of our repositories with the appropriate templates, workflows and actions.

Side note; we’ve been investigating a setup where we make changes to one repository and it then disperses to all the other repositories. It’s not pretty, but it’s an interim setup until we get submodule support

**Reply by chrispat** [reply]
To be completely transparent, given there are other features like:

- Dependabot config
- PR and Issue templates
- config for some GitHub apps
- unknown assumptions by other integrations

Supporting the .github folder being a submodule is a large effort beyond just actions.  That along with other priorities make it unlikely to happen.

**Reply by FaiziAhmi** [reply]
This code will help you as it will also help me a lot.
name: Submodule Update

on:
  push:
    paths:
      - 'path/to/submodule/**'

jobs:
  update:
    runs-on: ubuntu-latest

    steps:
      - name: Check out code
        uses: actions/checkout@v2

      - name: Update submodules
        run: |
          git submodule update --init --recursive
[.](https://ehsastracking.pk)

**Comment by backmeupplz** [comment]
+1 for the submodules to reuse the `.github` configs. See github.com/BigWhaleLabs repositories — we keep reusing the same `.github` folder across a couple dozen of repos, this is just unmainainable :(

**Comment by Forestsoft-de** [comment]
+1 same here. special actions (release, pr createion etc) on special conditions needs to be redefined in every process. even its the same for all repos.

**Comment by inson1** [comment]
reusable templates would be great, is there any update on this?

**Comment by renakdup** [comment]
+1, I have the same issue. Would be great to have the ability to share GitHub actions configs across repositories

**Reply by Leon-Africa** [reply]
+1

**Comment by IreshMM** [comment]
+1 here as well. It would be great to share .github directory across multiple repositories and track it independently.

**Comment by creatifcoding** [comment]
+1. suprised this isn't the default design. git submodules is such a counter-intuitive porcelain.

**Comment by DKorablin** [comment]
Hey.
An alternative approach can be used via GitHub Action.
[Github Workflow Sync](https://github.com/marketplace/actions/github-workflow-sync)

Shared files will be automatically copied to the required repositories.
